File: backend-go/internal/domain/engine/versioning.go
package engine

import (
	"fmt"
	"sort"

	"github.com/tiagofur/muebles-backend/internal/domain"
)

// DefaultStructureRevision is the assumed revision for legacy structures that
// omit Revision (#108). Mirrors TS DEFAULT_STRUCTURE_REVISION.
const DefaultStructureRevision = 1

// StructureRevisionError is a domain error raised by ResolveStructureRevision
// when a pin points to a revision that is neither the current one nor present
// in History. Mirrors TS ResolutionError context.
type StructureRevisionError struct {
	StructureID        string
	StructureCode      string
	Pin                int
	CurrentRevision    int
	AvailableRevisions []int
}

func (e *StructureRevisionError) Error() string {
	return fmt.Sprintf(
		"structure revision pin %d not found for structure %q (current=%d, available=%v)",
		e.Pin, e.StructureCode, e.CurrentRevision, e.AvailableRevisions,
	)
}

// StructureRevisionNumber normalises a Structure.Revision value, treating 0
// (legacy / missing) as DefaultStructureRevision. Mirrors TS structureRevision().
func StructureRevisionNumber(s domain.Structure) int {
	if s.Revision <= 0 {
		return DefaultStructureRevision
	}
	return s.Revision
}

// SnapshotStructureRevision builds an immutable StructureRevision from the
// current BOM-relevant fields of s. Mirrors TS snapshotStructureRevision().
func SnapshotStructureRevision(s domain.Structure) domain.StructureRevision {
	return domain.StructureRevision{
		Revision:   StructureRevisionNumber(s),
		Code:       s.Code,
		Name:       s.Name,
		WidthMm:    s.WidthMm,
		HeightMm:   s.HeightMm,
		DepthMm:    s.DepthMm,
		Components: s.Components,
		Presets:    s.Presets,
	}
}

// BumpStructureRevision applies an edit to a published structure (#108):
// the previous revision is snapshotted and prepended to History, and the
// returned structure carries Revision = current+1 plus the new draft fields.
//
// Mirrors TS bumpStructureRevision(). The caller passes nextDraft already
// populated with the new editable fields; ID/Revision/History are derived here
// and override anything nextDraft carries (matches the TS smuggling guard).
func BumpStructureRevision(current domain.Structure, nextDraft domain.Structure) domain.Structure {
	oldRevision := StructureRevisionNumber(current)
	snapshot := SnapshotStructureRevision(current)
	existingHistory := current.History
	if existingHistory == nil {
		existingHistory = []domain.StructureRevision{}
	}
	history := make([]domain.StructureRevision, 0, len(existingHistory)+1)
	history = append(history, snapshot)
	history = append(history, existingHistory...)

	return domain.Structure{
		ID:         current.ID,
		Code:       nextDraft.Code,
		Name:       nextDraft.Name,
		WidthMm:    nextDraft.WidthMm,
		HeightMm:   nextDraft.HeightMm,
		DepthMm:    nextDraft.DepthMm,
		Components: nextDraft.Components,
		Presets:    nextDraft.Presets,
		Notes:      nextDraft.Notes,
		Active:     nextDraft.Active,
		Revision:   oldRevision + 1,
		History:    history,
		CreatedAt:  current.CreatedAt,
		UpdatedAt:  current.UpdatedAt,
	}
}

// ResolveStructureRevision resolves which revision of s to use for BOM
// resolution, honoring pin (#108):
//   - pin == nil or pin == current → current.
//   - pin matches an entry in History → that frozen snapshot.
//   - otherwise → *StructureRevisionError (never silent fallback).
//
// Mirrors TS resolveStructureRevision() including the rich error context.
func ResolveStructureRevision(s domain.Structure, pin *int) (domain.StructureRevision, error) {
	current := StructureRevisionNumber(s)

	if pin == nil || *pin == current {
		return domain.StructureRevision{
			Revision:   current,
			Code:       s.Code,
			Name:       s.Name,
			WidthMm:    s.WidthMm,
			HeightMm:   s.HeightMm,
			DepthMm:    s.DepthMm,
			Components: s.Components,
			Presets:    s.Presets,
		}, nil
	}

	for _, h := range s.History {
		if h.Revision == *pin {
			return h, nil
		}
	}

	available := []int{current}
	for _, h := range s.History {
		available = append(available, h.Revision)
	}
	sort.Ints(available)

	return domain.StructureRevision{}, &StructureRevisionError{
		StructureID:        s.ID,
		StructureCode:      s.Code,
		Pin:                *pin,
		CurrentRevision:    current,
		AvailableRevisions: available,
	}
}

// ResolveStructureForPin resolves s honoring pin and returns a fully-shaped
// domain.Structure ready to feed ResolveBom / expandComposedModuleParts.
//
// ID/Notes/Active/History/CreatedAt/UpdatedAt are carried over from s so the
// object is shape-complete; BOM-relevant fields come from the resolved
// revision. Mirrors TS resolveStructureForPin().
func ResolveStructureForPin(s domain.Structure, pin *int) (domain.Structure, error) {
	resolved, err := ResolveStructureRevision(s, pin)
	if err != nil {
		return domain.Structure{}, err
	}
	out := s
	out.Code = resolved.Code
	out.Name = resolved.Name
	out.WidthMm = resolved.WidthMm
	out.HeightMm = resolved.HeightMm
	out.DepthMm = resolved.DepthMm
	out.Components = resolved.Components
	out.Presets = resolved.Presets
	out.Revision = resolved.Revision
	return out, nil
}

// CaptureProjectItemStructurePins pegs StructureRevisionPin onto each project
// item based on its module's current structure revision (#108). Used when a
// project is closed (quoted/accepted/produced) so the BOM stays frozen against
// later edits. Reopening conserves pins (mirrors TS reopen semantics).
//
// Items whose module has no structure or whose structure is missing from the
// catalog are returned unchanged (pin set to nil).
//
// Mirrors TS captureProjectItemStructurePins().
func CaptureProjectItemStructurePins(items []domain.ProjectItem, catalog domain.Catalog) []domain.ProjectItem {
	modulesByID := make(map[string]domain.Module, len(catalog.Modules))
	for _, m := range catalog.Modules {
		modulesByID[m.ID] = m
	}
	structuresByID := make(map[string]domain.Structure, len(catalog.Structures))
	for _, s := range catalog.Structures {
		structuresByID[s.ID] = s
	}

	out := make([]domain.ProjectItem, len(items))
	for i, item := range items {
		module, ok := modulesByID[item.ModuleID]
		if !ok || module.StructureID == "" {
			out[i] = item
			continue
		}
		structure, ok := structuresByID[module.StructureID]
		if !ok {
			out[i] = item
			continue
		}
		rev := StructureRevisionNumber(structure)
		// Always write a fresh pin (overwrites stale ones from previous closes).
		pinned := item
		pinned.StructureRevisionPin = &rev
		out[i] = pinned
	}
	return out
}

File: packages/domain/src/structures/versioning.ts
/**
 * Structure revision versioning (#108 / Slice 1).
 *
 * A published `Structure` is versioned: each edit produces a new `revision`
 * (monotonic) and the previous revision's BOM-relevant fields are pushed as an
 * immutable snapshot onto `history`. When a project is closed, each line item
 * captures the current `structureRevisionPin` so the frozen BOM can be
 * re-resolved against the exact revision the quote used — even after the
 * catalog structure is edited, replaced, or soft-deleted.
 *
 * Boundary: pure domain logic, no I/O. Mirrors the `QuotePriceSnapshot`
 * precedent (engine.ts `captureQuoteSnapshot`).
 */

import type {
  Catalog,
  ProjectItem,
  Structure,
  StructureRevision,
} from '../types';
import { ResolutionError } from '../errors';

/** Default revision assumed for legacy structures that omit `revision` (#108). */
export const DEFAULT_STRUCTURE_REVISION = 1;

/**
 * Normalized revision number — `undefined` (legacy data) is treated as 1.
 * Exposed for callers that need to compare revisions without the normalization dance.
 */
export function structureRevision(structure: Structure): number {
  return structure.revision ?? DEFAULT_STRUCTURE_REVISION;
}

/**
 * Build the immutable snapshot of the *current* structure's BOM-relevant fields.
 * Only fields that affect `resolveBom` are captured (notes/active are irrelevant
 * for re-resolution and intentionally dropped).
 */
export function snapshotStructureRevision(
  structure: Structure,
): StructureRevision {
  return {
    revision: structureRevision(structure),
    code: structure.code,
    name: structure.name,
    externalDims: structure.externalDims,
    presets: structure.presets,
    components: structure.components,
  };
}

/**
 * Apply an edit to a published structure (#108): the previous revision is
 * snapshotted and prepended to `history`, and the returned structure carries
 * `revision = current.revision + 1` plus the new draft fields.
 *
 * `nextDraft` carries the editable fields (everything except `id`, `revision`,
 * `history` — those are derived). Snapshots are prepended (newest-first) so
 * `history[0]` is always the most recently superseded revision.
 *
 * Returns the new structure and the previous revision number (handy for audits).
 */
export function bumpStructureRevision(
  current: Structure,
  nextDraft: Omit<Structure, 'id' | 'revision' | 'history'>,
): { structure: Structure; oldRevision: number } {
  const oldRevision = structureRevision(current);
  const snapshot = snapshotStructureRevision(current);
  const structure: Structure = {
    ...nextDraft,
    id: current.id,
    revision: oldRevision + 1,
    // Prepend newest-first; existing history (if any) follows in order.
    history: [snapshot, ...(current.history ?? [])],
  };
  return { structure, oldRevision };
}

/** Fields needed to re-resolve a pinned revision exactly as it was. */
export interface ResolvedStructureRevision {
  readonly revision: number;
  readonly code: string;
  readonly name: string;
  readonly externalDims?: Structure['externalDims'];
  readonly presets?: Structure['presets'];
  readonly components?: Structure['components'];
}

/**
 * Resolve which revision of `structure` should be used for BOM resolution,
 * honoring a pinned revision (#108):
 *
 * - `pin === undefined` → current revision (live).
 * - `pin === current revision` → current revision (live).
 * - `pin` matches an entry in `history` → that frozen snapshot.
 * - otherwise → throw `ResolutionError`. This covers the degenerate case where
 *   the structure was deleted without leaving a snapshot, or the pin points at
 *   a revision that was never snapshotted. We fail loudly with context rather
 *   than silently falling back to live (which would corrupt a frozen quote).
 *
 * `current` is accepted as `Structure` for ergonomic chaining; callers that only
 * have a partial structure should build a stub. The `id` is echoed back when
 * provided so error context can reference it.
 */
export function resolveStructureRevision(
  structure: Structure,
  pin?: number,
): ResolvedStructureRevision {
  const currentRevision = structureRevision(structure);

  if (pin === undefined || pin === currentRevision) {
    return {
      revision: currentRevision,
      code: structure.code,
      name: structure.name,
      externalDims: structure.externalDims,
      presets: structure.presets,
      components: structure.components,
    };
  }

  const matched = structure.history?.find((h) => h.revision === pin);
  if (matched) {
    return {
      revision: matched.revision,
      code: matched.code,
      name: matched.name,
      externalDims: matched.externalDims,
      presets: matched.presets,
      components: matched.components,
    };
  }

  const availableRevisions = [
    currentRevision,
    ...(structure.history ?? []).map((h) => h.revision),
  ].sort((a, b) => a - b);

  throw new ResolutionError(
    `Structure revision pin ${pin} not found for structure "${structure.code}"`,
    {
      structureId: structure.id,
      structureCode: structure.code,
      pin,
      currentRevision,
      availableRevisions,
      field: 'structureRevisionPin',
    },
  );
}

/**
 * Reify a resolved revision back into a full `Structure` so it can be fed to
 * `resolveComposedModule` / `resolveBom` unchanged. `id`/`notes`/`active`/
 * `history`/`revision` are carried over from `structure` since they don't
 * affect BOM resolution but keep the object shape-complete.
 *
 * Used by `resolveBom` when a pin is present.
 */
export function reifyResolvedStructure(
  structure: Structure,
  resolved: ResolvedStructureRevision,
): Structure {
  return {
    ...structure,
    code: resolved.code,
    name: resolved.name,
    externalDims: resolved.externalDims,
    presets: resolved.presets,
    components: resolved.components,
    revision: resolved.revision,
  };
}

/**
 * Convenience: resolve `structure` honoring `pin` and return a full `Structure`
 * ready to pass into BOM resolution. Throws via `resolveStructureRevision` for
 * unknown pins.
 */
export function resolveStructureForPin(
  structure: Structure,
  pin?: number,
): Structure {
  const resolved = resolveStructureRevision(structure, pin);
  return reifyResolvedStructure(structure, resolved);
}

/**
 * Peg `structureRevisionPin` onto each project item based on its module's
 * current structure revision (#108). Used when a project is closed
 * (quoted/accepted/produced) so the BOM stays frozen against later edits.
 *
 * Items whose module has no structure (or whose structure is missing from the
 * catalog) are returned unchanged — there is nothing to pin.
 */
export function captureProjectItemStructurePins(
  items: readonly ProjectItem[],
  catalog: Catalog,
): readonly ProjectItem[] {
  const modulesById = new Map(catalog.modules.map((m) => [m.id, m]));
  const structuresById = new Map(
    (catalog.structures ?? []).map((s) => [s.id, s]),
  );

  return items.map((item) => {
    const module = modulesById.get(item.moduleId);
    if (!module?.structureId) return item;
    const structure = structuresById.get(module.structureId);
    if (!structure) return item;
    return { ...item, structureRevisionPin: structureRevision(structure) };
  });
}
